sdk/include/platform/sunburst/platform-i2c.hh

```diff
@@ -0,0 +1,385 @@
+		  round_up_divide(SpcThigh[mode], clkPeriod);              // Spec. min.
+		uint16_t tLow = round_up_divide(SpcTlow[mode], clkPeriod); // Spec. min.
+		uint16_t tFall =
+		  round_up_divide(20 * 3 / 5, clkPeriod); // Spec. min. 3.3V
+		uint16_t tRise = round_up_divide(120, clkPeriod);
+		// Setup and Hold times for Start
+		uint16_t thdSta = round_up_divide(SpcThdSta[mode], clkPeriod);
+		uint16_t tsuSta = round_up_divide(SpcTsuSta[mode], clkPeriod);
+		// Setup and Hold times for Data
+		uint16_t thdDat = round_up_divide(SpcThdDat[mode], clkPeriod);
+		uint16_t tsuDat = round_up_divide(SpcTsuDat[mode], clkPeriod);
+		uint16_t tBuf   = round_up_divide(SpcTBuf[mode], clkPeriod);
+		uint16_t tsuSto = round_up_divide(SpcTsuSto[mode], clkPeriod);
+
+		// Prevent counters underflowing
+		if (tLow < thdDat + 1u)
+			tLow = thdDat + 1u;
+		if (tBuf < tsuSta + 1u)
+			tBuf = tsuSta + 1u;
+
+		timing0 = (tLow << 16) | tHigh;
+		timing1 = (tFall << 16) | tRise;
+		timing2 = (thdSta << 16) | tsuSta;
+		timing3 = (thdDat << 16) | tsuDat;
+		timing4 = (tBuf << 16) | tsuSto;
+	}
+
+	void blocking_write_byte(const uint32_t Fmt) volatile
+	{
+		while (0 != (StatusFormatFull & status)) {}
+		formatData = Fmt;
+	}
+
+	/// Returns true when the format fifo is empty
+	[[nodiscard]] bool format_empty() volatile
+	{
+		return 0 != (StatusFormatEmpty & status);
+	}
+
+	void blocking_write(const uint8_t  Addr7,
+	                    const uint8_t  data[],
+	                    const uint32_t NumBytes,
+	                    const bool     SkipStop) volatile
+	{
+		if (NumBytes == 0)
+			return;
+		blocking_write_byte(FormatDataStart | (Addr7 << 1) | 0u);
+		for (uint32_t i = 0; i < NumBytes - 1; ++i)
+		{
+			blocking_write_byte(data[i]);
+		}
+		blocking_write_byte((SkipStop ? 0u : FormatDataStop) |
+		                    data[NumBytes - 1]);
+	}
+
+	[[nodiscard]] bool blocking_read(const uint8_t  Addr7,
+	                                 uint8_t        buf[],
+	                                 const uint32_t NumBytes) volatile
+	{
+		for (uint32_t idx = 0; idx < NumBytes; idx += UINT8_MAX)
+		{
+			blocking_write_byte(FormatDataStart | (Addr7 << 1) | 1u);
+			while (!format_empty()) {}
+			if (interrupt_asserted(OpenTitanI2cInterrupt::Nak))
+			{
+				clear_interrupt(OpenTitanI2cInterrupt::Nak);
+				return false;
+			}
+			uint32_t bytesRemaining = NumBytes - idx;
+			bool     lastChunk      = UINT8_MAX >= bytesRemaining;
+			uint8_t  chunkSize =
+              lastChunk ? static_cast<uint8_t>(bytesRemaining) : UINT8_MAX;
+
+			blocking_write_byte((lastChunk ? FormatDataStop : 0) |
+			                    FormatDataReadBytes | chunkSize);
+			while (!format_empty()) {}
+
+			for (uint32_t chunkIdx = 0; chunkIdx < chunkSize; ++chunkIdx)
+			{
+				buf[idx + chunkIdx] = readData;
+			}
+		}
+		return true;
+	}
+
+	/// Returns true if the given interrupt is asserted.
+	[[nodiscard]] bool
+	interrupt_asserted(const OpenTitanI2cInterrupt Interrupt) volatile
```

Comment:
Consider `interrupt_is_asserted`.

Avoid `const` for value parameters.

This could also do with some explanation of how these relate to interrupts in the interrupt controller.

Reply:
I tend to use `const` by default even for value parameters, to ensure I don't modify variables I don't intend to. What are the disadvantages of having the extra `const` there?

(I have changed these to remove the `const` .)

Reply:
Const in a parameter is an API contract where you promise that you won’t mutate it. Const for a value type is not a meaningful contract because it is not observable to the caller whether you mutate a non-aliased object. It increases cognitive load on the reader because you have written something that looks like an API contract but is not.  If you then do need to mutate it, it either complicates the code by making you have two live values that a reader needs to track (even though one is not changed) or requires you to change the function signature (which can have further knock-on effects on overload resolution, particularly in the presence of subclasses).